### ISSUE TITLE
bug: no-pass-live-state-to-parent and no-pass-data-to-parent triggering with lodash debounce function

### ISSUE THREAD

**Issue by simarmol** [NickvanDyke/eslint-plugin-react-you-might-not-need-an-effect#62]
### Did you check docs and existing issues?

- [x] I have read the plugin docs
- [x] I have searched the existing issues
- [x] I have read [You Might Not Need An Effect](https://react.dev/learn/you-might-not-need-an-effect) and am reasonably confident that I *do* need an effect

### Plugin version

0.9.1

### ESLint version

9.39.1

### Describe the bug

I had a much larger hook I was getting this error in. I made this minimal chunk to reproduce the issue. I don't understand why the rules are triggering in this case. I'm not sure what it's considering the parent.

### Steps To Reproduce

```ts
import { useState, useEffect } from 'react';
import debounce from 'lodash/debounce';

export const useDebouncedState = (delay: number) => {
  const [value] = useState(0);

  const debouncedFunction = debounce((newValue) => {
    console.log(newValue);
  }, delay);

  useEffect(() => {
    debouncedFunction(value);
  }, [value, debouncedFunction]);

  return [];
};
```

I get these two errors:

```
  12:5  error  Avoid passing live state to parents in an effect. Instead, lift the state to the parent and pass it down to the child as a prop    react-you-might-not-need-an-effect/no-pass-live-state-to-parent
  12:5  error  Avoid passing data to parents in an effect. Instead, let the parent fetch the data itself and pass it down to the child as a prop  react-you-might-not-need-an-effect/no-pass-data-to-parent
```

If I delete the delay from the debounced function I have no errors.

```ts
import { useState, useEffect } from 'react';
import debounce from 'lodash/debounce';

export const useDebouncedState = (delay: number) => {
  const [value] = useState(0);

  const debouncedFunction = debounce((newValue) => {
    console.log(newValue);
  });

  useEffect(() => {
    debouncedFunction(value);
  }, [value, debouncedFunction]);

  return [];
};
```

If I only delete the console.log, I get `no-pass-live-state-to-parent` only.

```ts
import { useState, useEffect } from 'react';
import debounce from 'lodash/debounce';

export const useDebouncedState = (delay: number) => {
  const [value] = useState(0);

  const debouncedFunction = debounce((newValue) => {}, delay);

  useEffect(() => {
    debouncedFunction(value);
  }, [value, debouncedFunction]);

  return [];
};

```

```
10:5   error    Avoid passing live state to parents in an effect. Instead, lift the state to the parent and pass it down to the child as a prop  react-you-might-not-need-an-effect/no-pass-live-state-to-parent
```

### Expected Behavior

I don't believe any of the code snippets I put in the repro steps should be triggering `no-pass-live-state-to-parent` or `no-pass-data-to-parent`.

**Comment by nickjvandyke** [comment]
Thank you for the report and investigating and test cases! This definitely seems sub-optimal, and I think is a compromise in the heuristic.

I believe `Avoid passing data to parents` flags because you pass `delay` (an argument to your hook, and a prop if this were a component) to `debounce`. We can't (generally) know what happens inside `debounce`, so we assume worst-case: `value` is a callback (note we don't have type information) that `debounce` could internally pass data to.

Similar story for `Avoid passing live state to parent`: you pass `value` (state) to `debounce` and thus also `console.log`. Same as above, we can't know what those external functions do, so we assume the worst-case.

What's your intended functionality here? These error messages are definitely misleading, but I think an effect may indeed be unnecessary. You trigger the effect when `value` changes, thus you could instead call `debouncedFunction` wherever you call `setValue`. I could be missing context though. Just trying to clarify things so I don't introduce another bug trying to fix this one haha.

For this specific case, we could skip analyzing when we're in a custom hook. But I think this issue could still occur in components.

Maybe we assume best-case instead? That is, ignore when props/args are passed to another function - only analyze them when they themselves are directly called. I don't think we can get a perfectly accurate heuristic here, but that may be the lesser of two evils. Generally I would prefer some false negatives over some false positives.

I'm just thinking out loud here, but feel free to offer any input you might have 🙂

**Comment by simarmol** [comment]
Got the go-ahead for posting the actual component:
```tsx
export function useDebouncedState<T>(
  initialValue: T | (() => T),
  delay?: number,
  options: UseDebouncedStateOptions = {},
): [T, T, React.Dispatch<React.SetStateAction<T>>, () => void] {
  const [value, setValue] = useState<T>(initialValue);
  const [debouncedValue, setDebouncedValue] = useState<T>(initialValue);
  const { leading = false, maxWait = null, trailing = true } = options;
  const isFirstRenderRef = useRef(true);

  const debouncedSetValue = useMemo(() => {
    return debounce(
      (newValue: T) => {
        setDebouncedValue(newValue);
      },
      delay ?? 500,
      { leading, trailing, ...(maxWait !== null && { maxWait }) },
    );
  }, [delay, leading, maxWait, trailing]);

  const flush = useCallback(() => {
    debouncedSetValue.flush();
    setDebouncedValue(value);
  }, [debouncedSetValue, value]);

  useEffect(() => {
    if (isFirstRenderRef.current) {
      // values already match on the first render, so don't start
      // any timers. This makes a leading edge option able to
      // immediately receive a value.
      isFirstRenderRef.current = false;
      return;
    }

    debouncedSetValue(value);
  }, [value, debouncedSetValue]);

  useEffect(() => {
    return () => {
      debouncedSetValue.cancel();
    };
  }, [debouncedSetValue]);

  return [value, debouncedValue, setValue, flush];
}
```
